src/Functions/PartitionByHyperplanes.cpp

```diff
@@ -0,0 +1,338 @@
+
+    bool isVariadic() const override { return true; }
+    bool isSuitableForShortCircuitArgumentsExecution(const DataTypesWithConstInfo & /*arguments*/) const override { return true; }
+
+    size_t getNumberOfArguments() const override { return 0; }
+    bool useDefaultImplementationForConstants() const override { return true; }
+
+    DataTypePtr getReturnTypeImpl(const ColumnsWithTypeAndName & arguments) const override
+    {
+        FunctionArgumentDescriptors mandatory_args{
+            {"vectors", &isArray<IDataType>, nullptr, "Array"},
+            {"normals", &isArray<IDataType>, nullptr, "Array"},
+        };
+
+        FunctionArgumentDescriptors optional_args{
+            {"offsets", &isArray<IDataType>, nullptr, "Array"},
+        };
+
+        validateFunctionArgumentTypes(*this, arguments, mandatory_args, optional_args);
+
+        return std::make_shared<DataTypeString>();
+    }
+
+    ColumnPtr executeImpl(const ColumnsWithTypeAndName & arguments, const DataTypePtr & result_type, size_t input_rows_count) const override
+    {
+        const ColumnArray * vectors = typeid_cast<const ColumnArray *>(arguments[0].column.get());
+        if (!vectors)
+            throw Exception(ErrorCodes::ILLEGAL_COLUMN, "First argument of function {} must be Array", getName());
+        const auto & nested_vectors_data = typeid_cast<const ColumnFloat32 &>(vectors->getData());
+
+        const ColumnArray * normals = typeid_cast<const ColumnArray *>(arguments[1].column.get());
+        if (!normals)
+            throw Exception(ErrorCodes::ILLEGAL_COLUMN, "Second argument of function {} must be Array", getName());
+        const auto & nested_normals_data = typeid_cast<const ColumnFloat32 &>(normals->getData());
+
+        const size_t vector_size = vectors->getOffsets().front();
+
+        auto const_offsets = ColumnConst::create(ColumnFloat32::create(1, 0), input_rows_count * vector_size);
+        const IColumn * nested_offsets_data = const_offsets.get();
+        if (arguments.size() >= 3)
+        {
+            const ColumnArray * offsets = typeid_cast<const ColumnArray *>(arguments[2].column.get());
+            if (!offsets)
+                throw Exception(ErrorCodes::ILLEGAL_COLUMN, "Third argument of function {} must be Array", getName());
+            nested_offsets_data = &offsets->getData();
+        }
+
+        if (input_rows_count == 0)
+            return result_type->createColumn();
+
+        checkDimension(vectors->getOffsets(), vector_size);
+        checkDimension(normals->getOffsets(), vector_size);
+
+        auto col_res = ColumnFloat32::create(input_rows_count * input_rows_count);
+        auto col_res_offsets = ColumnArray::ColumnOffsets::create();
+        col_res_offsets->reserve(input_rows_count);
+        for (size_t i = 0; i < input_rows_count; ++i)
+        {
+            col_res_offsets->insertValue(input_rows_count * i);
+        }
+
+        executeInternal(
```

Comment:
Seems `ColumnsType` template argument always `ColumnFloat32`, so we can get rid of it, supporting only floats should be ok. We have to decide Float32/64, though.